pkg/cmd/env/options.go

```diff
@@ -0,0 +1,281 @@
+/*
+SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package env
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"runtime"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/gardener/gardenctl-v2/internal/gardenclient"
+	"github.com/gardener/gardenctl-v2/internal/util"
+	"github.com/gardener/gardenctl-v2/pkg/cmd/base"
+	"github.com/gardener/gardenctl-v2/pkg/target"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+)
+
+type options struct {
+	base.Options
+
+	// Unset resets environment variables and configuration of the cloudprovider CLI for your shell.
+	Unset bool
+	// Shell to configure.
+	Shell string
+	// GardenDir is the configuration directory of gardenctl.
+	GardenDir string
+	// CmdPath is the path of the called command.
+	CmdPath string
+	// CurrentTarget is the current target
+	CurrentTarget target.Target
+	// ProviderType is the name of the cloud provider
+	ProviderType string
+	// Template is the script template
+	Template Template
+}
+
+// Complete adapts from the command line args to the data required.
+func (o *options) Complete(f util.Factory, cmd *cobra.Command, args []string) error {
+	o.Shell = cmd.Name()
+	o.CmdPath = cmd.Parent().CommandPath()
+	o.GardenDir = f.GardenHomeDir()
+
+	switch o.ProviderType {
+	case "kubernetes":
+		o.Template = newTemplate("usage-hint", "templates/kubernetes.tmpl")
+	default:
+		o.Template = newTemplate("usage-hint")
+	}
+
+	return nil
+}
+
+// Validate validates the provided command options.
+func (o *options) Validate() error {
+	if o.Shell == "" {
+		return pflag.ErrHelp
+	}
+
+	s := Shell(o.Shell)
+	if err := s.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AddFlags binds the command options to a given flagset.
+func (o *options) AddFlags(flags *pflag.FlagSet) {
+	var text string
+
+	switch o.ProviderType {
+	case "kubernetes":
+		text = "the KUBECONFIG environment variable"
+	default:
+		text = "the cloud provider CLI environment variables and logout"
+	}
+
+	usage := fmt.Sprintf("Generate the script to unset %s for %s", text, o.Shell)
+	flags.BoolVarP(&o.Unset, "unset", "u", o.Unset, usage)
+}
+
+// Run does the actual work of the command.
+func (o *options) Run(f util.Factory) error {
+	manager, err := f.Manager()
+	if err != nil {
+		return err
+	}
+
+	o.CurrentTarget, err = manager.CurrentTarget()
+	if err != nil {
+		return err
+	}
+
+	if o.CurrentTarget.GardenName() == "" {
+		return target.ErrNoGardenTargeted
+	}
+
+	switch o.ProviderType {
+	case "kubernetes":
+		return o.runKubernetes(f.Context(), manager)
+	default:
+		t := o.CurrentTarget
+		if t.ShootName() == "" {
+			return target.ErrNoShootTargeted
+		} else if t.ProjectName() == "" && t.SeedName() == "" {
+			return target.ErrNeitherProjectNorSeedTargeted
```

Comment:
Yes. I think, it should be allowed. I think we should also remove the check for both project and seed targeted at same time. Because it will work if the seed is correct otherwise it give an error anyway.

Reply:
✔️